### PR TITLE
BLD: remove distutils

### PR DESCRIPTION
**Pull request by fangchenli**
distutils will be removed in Python 3.12. This PR removes all of its usage from setup.py.

### CODE REVIEW

**Comment by gmarkall** [comment]
Many thanks for the PR. Why does it remove the `build_doc` command?

**Comment by gmarkall** [comment]
The issue that RTD seems to have also seems to crop up for me when building locally:

```
$ python setup.py develop
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/setup.py", line 75, in <module>
    class NumbaBuildExt(build_ext):
  File "/home/gmarkall/numbadev/numba/setup.py", line 77, in NumbaBuildExt
    user_options = build_ext.user_options + numba_be_user_options
AttributeError: module 'setuptools.command.build_ext' has no attribute 'user_options'
```

**Comment by fangchenli** [comment]
> Many thanks for the PR. Why does it remove the `build_doc` command?

It's not used in the codebase. I could add it back if people still want it.

File: setup.py
Comment (gmarkall) [inline setup.py:70]: I think we don't have these on in all cases and instead guard them behind the `-wall` and `-werror` flags because some users might have a newer C compiler than we use / test with, which triggers unexpected warnings and causes the build to fail.

**Comment by gmarkall** [comment]
> It's not used in the codebase. I could add it back if people still want it.

Thanks for clarifying. I would imagine we can probably do without it, but I don't know if any downstream packagers are using it to build docs as part of their packaging setup (e.g. in Debian, Fedora, etc.). Another maintainer might be able to provide more clarity here.

**Comment by gmarkall** [comment]
> > It's not used in the codebase. I could add it back if people still want it.
> 
> Thanks for clarifying. I would imagine we can probably do without it, but I don't know if any downstream packagers are using it to build docs as part of their packaging setup (e.g. in Debian, Fedora, etc.). Another maintainer might be able to provide more clarity here.

To follow on a little more thought here, I certainly think it would be easier to get over the line here if the PR just focused on removing distutils usage and not making other functional changes, then were followed up by any other cleanups / improvements (such as removing the doc build command).

**Comment by fangchenli** [comment]
> > > It's not used in the codebase. I could add it back if people still want it.
> > 
> > 
> > Thanks for clarifying. I would imagine we can probably do without it, but I don't know if any downstream packagers are using it to build docs as part of their packaging setup (e.g. in Debian, Fedora, etc.). Another maintainer might be able to provide more clarity here.
> 
> To follow on a little more thought here, I certainly think it would be easier to get over the line here if the PR just focused on removing distutils usage and not making other functional changes, then were followed up by any other cleanups / improvements (such as removing the doc build command).

Agree, I've put it back.

File: setup.py
Comment (gmarkall) [inline setup.py:80]: I think this is not the behaviour we want - we don't want these flags to always be added on Linux / x86_64 when we build, only on Linux / x86_64 in CI. This still removes the option for the user to add these flags through options to `build_ext` (and additionally the `noopt` flag as well..

**Comment by gmarkall** [comment]
In the present state, the `--noopt`, `--werror` and `--wall` flags are removed from `build_ext`, but they need to be retained. They are described in the documentation: https://numba.readthedocs.io/en/latest/developer/contributing.html#building-numba

**Comment by gmarkall** [comment]
Many thanks for the updates. Will have another look at this.

File: setup.py
Comment (gmarkall) [inline setup.py:76]: This came from the vendored Versioneer's `cmdclass`, which I think raises two questions:

- The vendored Versioneer we have is still using `distutils`. What's the right course of action here? Should we update the vendored version, or try to un-vendor it?
- Is `setuptools.command.build_ext.build_ext` the same thing as returned by `cmdclass.get('build_ext', build_ext)` when `cmdclass` is Versioneer's `cmdclass`?

(I also think there's the question of whether continuing to use Versioneer is the right way to go, but I *definitely* don't want to muddy the waters with that right now)

**Review by gmarkall — CHANGES_REQUESTED** [review]
Thanks for the updates - I've had a look and before proceeding I'd like to understand the Versioneer situation a bit better - see comments on the diff.

File: setup.py
Comment (fangchenli) [reply setup.py]: I reverted it. And will address versioneer issue later.

**Review by gmarkall — APPROVED** [review]
Many thanks for the updates. I think these changes look good. Considerations:

- The `spawn` command was just a wrapper around `subprocess.Popen`, so replacing it with `run` makes sense and makes things simpler.
- The `build` command is just a subclass of `Command`, but adds a whole load of behaviour completely unnecessary for the docs build, so making `build_doc` a `Command` is an improvement.

Other notes:

- The vendored Versioneer still uses distutils, but we should consider what to do about that separately from this PR.
- #8476 handles the other (non-setup.py) uses of distutils.
- This should maybe have a buildfarm run just in case of any surprises that might show up (@esc / @stuartarchibald thoughts maybe?) - marking as "Waiting on CI" for now before we decide to do that / do that.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cpu_yaml_166`.

**Comment by stuartarchibald** [comment]
> Buildfarm ID: `numba_smoketest_cpu_yaml_166`.

Passed.

**Pull request merged**
